src/lib.rs

```diff
@@ -355,21 +356,27 @@ impl fmt::Display for FormatReport {
 /// Format the given snippet. The snippet is expected to be *complete* code.
 /// When we cannot parse the given snippet, this function returns `None`.
 fn format_snippet(snippet: &str, config: &Config) -> Option<String> {
-    let mut out: Vec<u8> = Vec::with_capacity(snippet.len() * 2);
-    let input = Input::Text(snippet.into());
     let mut config = config.clone();
-    config.set().emit_mode(config::EmitMode::Stdout);
-    config.set().verbose(Verbosity::Quiet);
-    config.set().hide_parse_errors(true);
-    {
-        let mut session = Session::new(config, Some(&mut out));
-        let result = session.format(input);
-        let formatting_error = session.errors.has_macro_format_failure
-            || session.out.as_ref().unwrap().is_empty() && !snippet.is_empty();
-        if formatting_error || result.is_err() {
-            return None;
+    let out = panic::catch_unwind(|| {
+        let mut out: Vec<u8> = Vec::with_capacity(snippet.len() * 2);
+        config.set().emit_mode(config::EmitMode::Stdout);
+        config.set().verbose(Verbosity::Quiet);
+        config.set().hide_parse_errors(true);
+        let formatting_error = {
+            let input = Input::Text(snippet.into());
+            let mut session = Session::new(config, Some(&mut out));
+            let result = session.format(input);
+            session.errors.has_macro_format_failure
+                || session.out.as_ref().unwrap().is_empty() && !snippet.is_empty()
+                || result.is_err()
+        };
+        if formatting_error {
+            None
+        } else {
+            Some(out)
         }
-    }
+    })
+    .ok()??;
```

Comment:
a comment about the need for double question marks would be nice I think

Reply:
Agreed. It is indeed not clear from code point of view at first glance.

Reply:
Note that `panic::catch_unwind(|| { .. })` returns `Result<Option<String>>`. So if I write the expression in multiple lines, it would be like the following:

```rust
let out: Result<Option<String>> = panic::catch_unwind(|| { /* ... */ });
let out: Option<Option<String>> = out.ok();
let out: Option<String> = out?;
let out: String = out?;
```

So the first try operator handles the error from `catch_unwind`, whereas the second one handles `None` returned from the closure.

Reply:
I think @scampi suggests to add a short comment on the code of the reasoning for it.

Reply:
@topecongiro I meant what @otavio said, sorry for the lack of clarity. The last line of your https://github.com/rust-lang-nursery/rustfmt/pull/3103#discussion_r225216927 would be perfect: first `?` for the possible panic, second `?` for handling the `None`.